com.oracle.truffle.r.runtime/src/com/oracle/truffle/r/runtime/nodes/altrep/AltrepDuplicateNode.java

```diff
@@ -0,0 +1,75 @@
+package com.oracle.truffle.r.runtime.nodes.altrep;
+
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import com.oracle.truffle.r.runtime.data.NativeDataAccess.NativeMirror;
+import com.oracle.truffle.r.runtime.data.RIntArrayVectorData;
+import com.oracle.truffle.r.runtime.data.RIntVector;
+import com.oracle.truffle.r.runtime.data.altrep.AltIntegerClassDescriptor;
+import com.oracle.truffle.r.runtime.data.altrep.AltrepUtilities;
+import com.oracle.truffle.r.runtime.ffi.AltrepRFFI;
+import com.oracle.truffle.r.runtime.ffi.util.NativeMemory;
+import com.oracle.truffle.r.runtime.ffi.util.NativeMemory.ElementType;
+import com.oracle.truffle.r.runtime.nodes.RBaseNode;
+
+/**
+ * Note that the behavior of GNU-R's Rf_duplicate in duplicate.c is a bit weird: When the altrep vector
+ * does not have <code>Duplicate</code> method registered, then calling <code>Rf_duplicate</code> on such
+ * altrep vector creates an atomic vector, not an altrep vector.
+ */
+@GenerateUncached
+public abstract class AltrepDuplicateNode extends RBaseNode {
```

Comment:
This can be inlined as message into the `RAltIntVectorData`
```
@ExportMessage
static class MyMessage {
   @Specialization
   Object doIt(MyObject receiver, ...other message args...) { ...
   ...
}
```

Reply:
I know, I already tried it. I did it this way, because it seems more readable to me.